permissions/src/main/java/com/google/accompanist/permissions/PermissionState.kt

```diff
@@ -0,0 +1,208 @@
+        )
+    }
+}
+
+/**
+ * A state object that can be hoisted to control and observe permission status changes.
+ *
+ * In most cases, this will be created via [rememberPermissionState].
+ *
+ * @param permission the permission to control and observe.
+ * @param refreshShouldShowRationaleState action to refresh whether or not a rationale should be
+ * presented to the user to explain why the permission should be granted.
+ * @param launcher [ActivityResultLauncher] to ask for this permission to the user.
+ * @param hasPermissionState [State] that represents if the permission is granted.
+ * @param shouldShowRationaleState [State] that represents if the user should be presented with a
+ * rationale.
+ * @param permissionRequestedState [State] that represents if the permission has been requested
+ * previously.
+ */
+@Stable
+class PermissionState(
+    val permission: String,
+    private val refreshShouldShowRationaleState: () -> Unit,
+    private val launcher: ActivityResultLauncher<String>,
+    hasPermissionState: State<Boolean>,
+    shouldShowRationaleState: State<Boolean>,
+    permissionRequestedState: State<Boolean>
+) {
+    /**
+     * When `true`, the user has granted the [permission].
+     */
+    val hasPermission by hasPermissionState
+
+    /**
+     * When `true`, the user should be presented with a rationale.
+     */
+    val shouldShowRationale by shouldShowRationaleState
+
+    /**
+     * When `true`, the [permission] request has been done previously.
+     */
+    val permissionRequested by permissionRequestedState
+
+    /**
+     * Request the [permission] to the user.
+     *
+     * This triggers a system dialog that asks the user to grant or revoke the permission.
+     * Note that this dialog might not appear on the screen if the user doesn't want to be asked
+     * again or has denied the permission multiple times.
+     * This behavior varies depending on the Android level API.
+     */
+    fun launchPermissionRequest() = launcher.launch(permission)
+
+    /**
+     * Check if the rationale should be presented to the user.
+     *
+     * This triggers a state update in the [shouldShowRationale] property.
+     */
+    fun refreshShouldShowRationale() = refreshShouldShowRationaleState()
+}
+
+/**
+ * Creates a [MutableState] that represents if the user has granted the permission.
+ *
+ * This state is remembered across compositions and is updated every time the `lifecycle` of the
+ * current [LocalLifecycleOwner] receives the `ON_START` lifecycle event. This check is crucial
+ * when the user manually grants the permission on the Settings screen while the app is in the
+ * background. There's no need to check when the permission is revoked from the Settings screen
+ * as that triggers a process restart.
+ */
+@Composable
+private fun rememberPermissionGrantedState(permission: String): MutableState<Boolean> {
+    val context = LocalContext.current
+    val checkPermission = {
```

Comment:
Extract to a private function rather than a lambda? This will create an instance every time.

Reply:
Done! Thanks